cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -1240,11 +1250,141 @@ public long getProgramRunCount(ProgramId programId) {
     return result;
   }
 
+  void createStartTimeIfNotExist() {
+    MDSKey mdsKey = new MDSKey(RUN_COUNT_FIRST_UPGRADE_TIME);
+    if (get(mdsKey, Long.TYPE) == null) {
+      LOG.info("Start to upgrade the run count");
+      write(mdsKey, TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis()));
+    }
+  }
+
+  /**
+   * Compute the old run count if the run record is older than first start up time after 5.1
+   */
+  boolean computeOldRunCount(int maxRows) {
+    boolean upgradeDone = upgradeActiveRowkey(TYPE_RUN_RECORD_STARTING, maxRows);
+    upgradeDone &= upgradeActiveRowkey(TYPE_RUN_RECORD_STARTED, maxRows);
+    upgradeDone &= upgradeActiveRowkey(TYPE_RUN_RECORD_SUSPENDED, maxRows);
+    upgradeDone &= computeOldCompletedRunCount(maxRows);
+    return upgradeDone;
+  }
+
+  /**
+   * Upgrades the rowkeys for the active run record. Note that we do not count the run record of the old active runs
+   * since it is hard to deal with the races and we expect user to stop all the programs before they do the upgrade.
+   *
+   * @param recordType type of the record
+   * @param maxRows    maximum number of rows to be upgraded in this call.
+   * @return true if no rows required an upgrade
+   */
+  private boolean upgradeActiveRowkey(String recordType, int maxRows) {
+    LOG.info("Checking upgrade for {}", recordType);
```

Comment:
Upgrading run records in {} state

Reply:
It also seems like this message should be in the caller, as this method will be called multiple times.

Reply:
The `upgrade()` method in `DefaultStore` is called using a ScheduledService, so it will be hard to just log this once, as the `upgrade()` method itself will be called multiple times